Removed temporary strings when converting bytes to their ASCII hex representation.

mwinterb opened the pull request:
For large XML messages sent through SQS SendMessage, the time to UrlEncode the content can be significant.

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
My company pushes lots of XML through SQS's SendMessage. Since SendMessage takes the message URL encoded all of the XML syntactic "noise" needs to be encoded. UrlEncode shows up as a hot spot often in profiles.

There are most certainly ways to make this even faster, but I feel this is a good balance.

## Testing
<!--- Please describe in detail how you tested your changes -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I ran the tests in AWSSDK.CoreAndCustomUnitTests.NetStandard.sln, but unfortunately none of the tests called the changed code. As a smokescreen test, I extracted a 8K message (expands to 13K) and ran it through the old and new encoders to verify consistency in output, as well as verifying the performance increase with BenchmarkDotNet. The timing results are below.

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19045, VM=VirtualBox
11th Gen Intel Core i7-11700K 3.60GHz, 1 CPU, 6 logical and 6 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4614.0), X64 RyuJIT
  DefaultJob : .NET Framework 4.8 (4.8.4614.0), X64 RyuJIT


```
|       Method |     Mean |   Error |  StdDev |
|------------- |---------:|--------:|--------:|
| UrlEncodeAWS | 547.8 μs | 6.43 μs | 5.37 μs |
| UrlEncodeMDW | 242.8 μs | 1.18 μs | 1.05 μs |


## Screenshots (if appropriate)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

## License
<!--- The SDK is released under the [Apache 2.0 license][license], so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a [Contributor License Agreement (CLA)][cla] -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [x] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

normj commented:
I don't think this PR will have any noticeable performance gains for your use case that warrants the added complexity. The message being sent to SQS is not sent as a URL parameter and is instead part of the message body. The `ToHex` method is called a couple times during signing to hex encode a couple 32 byte hashes. Given all the work that has to be done especially through signing these 2 extra 64 length strings is going to be lost in the noise.

If you feel very strongly about this PR then the code needs a lot of comments for the choices it makes. The code is pretty cryptic for understandable reasons of maximizing performance but it stills needs to be understandable even for readers not deep into the mindset here.

mwinterb commented:
Hi Norm,

Thank you for your comments.

As an initial note: I'm adjacent to the team that uses the AWS SDK, so I can't justify this: They are still using 3.7.12. The SendMessageRequestMarshaller and this related code doesn't _seem_ like it's changed meaningfully, but if this is no longer an issue in current versions, I'll gladly pester them to update. Obviously if they were to get my proposed change they would need to update anyway, but being able to point to an obvious change makes it an easier sell than just "I dunno, upgrade, maybe it'll help".


> I don't think this PR will have any noticeable performance gains for your use case that warrants the added complexity. The message being sent to SQS is not sent as a URL parameter and is instead part of the message body. The `ToHex` method is called a couple times during signing to hex encode a couple 32 byte hashes. Given all the work that has to be done especially through signing these 2 extra 64 length strings is going to be lost in the noise.
> 
RE: SQS, you're right, I misunderstood. I just looked at the examples in the SendMessage docs and everything for the body uses application/x-www-form-urlencoded, so that's what lead me to think it was the URL parameter and not body. And also, just because the examples are x-www-form-urlencoded doesn't mean it's a requirement of the endpoint (I think? I'm not an HTTP person). But I'm still confused by the profile results I'm seeing. I don't know if I can share this ETL file, and I need to focus on something else for the next few days so I won't be able to step in to see exactly what's being passed to UrlEncode to confirm my suspicion that it's the message body causing problems. And I'll also verify this with a current release.

An incomplete summary of the CPU Usage (Sampled) trace:
13,194 samples for the relevant threads.
5,980 were for AWSSDKUtils.UrlEncode 
3,623 of those were from calls from AWS4Signer.GetRequestPayloadBytes
2,349 of those were from calls from HttpHandler.CreateWebRequest.

So ~55% of the samples are "us" and we need to fix that separately, but 45% being in UrlEncode seemed high and easily targeted.

Within UrlEncode, the heavy-hitters:
String.Format(IFormatProvider, System.String, System.Object[]): 4189
JIT_New: 615
JIT_NewArr1: 533
TextEncoding::GetBytes: 263
COMString::IndexOfChar: 227
StringBuilder::ToString: 72

JIT_NewArr1 and calling the System.Object[] overload are presumably because it targets .NET 4.5 instead of .NET 4.6. Convincing the group to switch to .NET 6 or later just for this possible fix would be more difficult. But since this change was still better when targeting 4.8, I don't know if the System.Object[] overload is really what's hurting.

For ToHex and EncodeTraceIdHeaderValue, those changes were just made for "consistency". I'm entirely uninterested in those and I think removing those may help make the final code slightly more clear on its own.
I've also written basic base16 encoders often, so it is something that is obvious to me, but I'll try to come up with decent comments that don't just end up reading like "set x to 5".

mwinterb commented:
Thank you for the poke.

I've reduced the changes to just UrlEncode. I think I've also corrected my misunderstanding of where the URL encoded format of the message to send gets used.

I took this sample code https://docs.aws.amazon.com/sdk-for-net/v3/developer-guide/SendMessage.html#SendMessage-complete-code
removed the usage of await just to make the callstacks clearer to follow, and, using the 3.7.200.1 packages for AWSSDK.Core, AWSSDK.SQS, and AWSSDK.SecurityToken on nuget.org that have today's date as the publishing date, I put a breakpoint on line 282 of `AWSSDKUtils.GetParametersAsString`.

During the `SendMessage` call for `JsonMessage`, the breakpoint was hit several times, two of which generated this string for result (before trimming off the end):
`Action=SendMessage&MessageBody=%7B%22product%22%3A%5B%7B%22name%22%3A%22Product%20A%22%2C%22price%22%3A%20%2232%22%7D%2C%7B%22name%22%3A%20%22Product%20B%22%2C%22price%22%3A%20%2227%22%7D%5D%7D&Version=2012-11-05&`

If you squint right, it's "clear" that the text after `MessageBody=` is the result of UrlEncode for `JsonMessage`.

One of the callstacks was this:
```
>	AWSSDK.Core.dll!Amazon.Util.AWSSDKUtils.GetParametersAsString(Amazon.Runtime.Internal.ParameterCollection parameterCollection) Line 282	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Auth.AWS4Signer.GetRequestPayloadBytes(Amazon.Runtime.Internal.IRequest request) Line 1044	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Auth.AWS4Signer.SetRequestBodyHash(Amazon.Runtime.Internal.IRequest request, bool signPayload, string chunkedBodyHash, int signatureLength) Line 544	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Auth.AWS4Signer.SignRequest(Amazon.Runtime.Internal.IRequest request, Amazon.Runtime.IClientConfig clientConfig, Amazon.Runtime.Internal.Util.RequestMetrics metrics, string awsAccessKeyId, string awsSecretAccessKey) Line 210	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Auth.AWS4Signer.Sign(Amazon.Runtime.Internal.IRequest request, Amazon.Runtime.IClientConfig clientConfig, Amazon.Runtime.Internal.Util.RequestMetrics metrics, string awsAccessKeyId, string awsSecretAccessKey) Line 120	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Auth.AWS4Signer.Sign(Amazon.Runtime.Internal.IRequest request, Amazon.Runtime.IClientConfig clientConfig, Amazon.Runtime.Internal.Util.RequestMetrics metrics, Amazon.Runtime.ImmutableCredentials credentials) Line 154	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Signer.SignRequest(Amazon.Runtime.IRequestContext requestContext) Line 143	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Signer.PreInvoke(Amazon.Runtime.IExecutionContext executionContext) Line 86	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Signer.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 41	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 44	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CredentialsRetriever.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 73	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.RetryHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 77	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CallbackHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 46	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.SQS.dll!Amazon.SQS.Internal.ProcessRequestHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 27	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.BaseEndpointResolver.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 34	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Marshaller.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 38	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CallbackHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 46	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.ErrorCallbackHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 39	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.MetricsHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 40	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.RuntimePipeline.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 136	C#
 	AWSSDK.Core.dll!Amazon.Runtime.AmazonServiceClient.Invoke<Amazon.SQS.Model.SendMessageResponse>(Amazon.Runtime.AmazonWebServiceRequest request, Amazon.Runtime.Internal.InvokeOptionsBase options) Line 209	C#
 	AWSSDK.SQS.dll!Amazon.SQS.AmazonSQSClient.SendMessage(Amazon.SQS.Model.SendMessageRequest request) Line 3555	C#
 	AWSSDK.SQS.dll!Amazon.SQS.AmazonSQSClient.SendMessage(string queueUrl, string messageBody) Line 3514	C#
 	SqsTest.exe!SQSSendMessages.Program.SendMessage(Amazon.SQS.IAmazonSQS sqsClient, string qUrl, string messageBody) Line 61	C#
 	SqsTest.exe!SQSSendMessages.Program.Main(string[] args) Line 41	C#
```

and the other was this:
```
>	AWSSDK.Core.dll!Amazon.Util.AWSSDKUtils.GetParametersAsString(Amazon.Runtime.Internal.ParameterCollection parameterCollection) Line 282	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.HttpHandler<System.IO.Stream>.CreateWebRequest(Amazon.Runtime.IRequestContext requestContext) Line 477	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.HttpHandler<System.IO.Stream>.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 67	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Unmarshaller.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 50	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.SQS.dll!Amazon.SQS.Internal.ValidationResponseHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 29	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.ErrorHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 72	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CallbackHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 46	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Signer.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 42	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.EndpointDiscoveryHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 44	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CredentialsRetriever.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 73	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.RetryHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 77	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CallbackHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 46	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.SQS.dll!Amazon.SQS.Internal.ProcessRequestHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 27	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.BaseEndpointResolver.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 34	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.Marshaller.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 38	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.CallbackHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 46	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.ErrorCallbackHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 39	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.PipelineHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 60	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.MetricsHandler.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 40	C#
 	AWSSDK.Core.dll!Amazon.Runtime.Internal.RuntimePipeline.InvokeSync(Amazon.Runtime.IExecutionContext executionContext) Line 136	C#
 	AWSSDK.Core.dll!Amazon.Runtime.AmazonServiceClient.Invoke<Amazon.SQS.Model.SendMessageResponse>(Amazon.Runtime.AmazonWebServiceRequest request, Amazon.Runtime.Internal.InvokeOptionsBase options) Line 209	C#
 	AWSSDK.SQS.dll!Amazon.SQS.AmazonSQSClient.SendMessage(Amazon.SQS.Model.SendMessageRequest request) Line 3555	C#
 	AWSSDK.SQS.dll!Amazon.SQS.AmazonSQSClient.SendMessage(string queueUrl, string messageBody) Line 3514	C#
 	SqsTest.exe!SQSSendMessages.Program.SendMessage(Amazon.SQS.IAmazonSQS sqsClient, string qUrl, string messageBody) Line 61	C#
 	SqsTest.exe!SQSSendMessages.Program.Main(string[] args) Line 41	C#
```

peterrsongg commented on sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:
```diff
@@ -1056,12 +1056,31 @@ public static string UrlEncode(int rfcNumber, string data, bool path)
+                    // Break apart the byte into two four-bit components and
+                    // then convert each into their hexadecimal equivalent.
+                    byte b = (byte)symbol;
+                    int hinibble = (b >> 4) & 0xF;
```
Wouldn't shifting 4 bits to the right make the bitwise AND operator unnecessary for the hinibble since we're isolating those bits anyways? I think you can remove the bitwise AND here.

peterrsongg commented:
This is just a general comment as well, but are those benchmarking numbers for URLEncode up to date with your latest revision?

ashishdhingra commented:
@mwinterb Please review the last comments and provide your inputs.

mwinterb commented:
The mask isn't necessary, removed.

Current profile results:
.NET Framework 4.8:

|                     Method |      Mean |    Error |   StdDev |
|--------------------------- |----------:|---------:|---------:|
|               UrlEncodeAWS | 597.33 us | 5.341 us | 4.996 us |
|               UrlEncodeMDW | 253.78 us | 1.275 us | 1.130 us |
|       UrlEncodeMDWWithMask | 255.69 us | 1.271 us | 1.189 us |
|  UrlEncodeAWS_SymbolSearch | 595.88 us | 7.325 us | 6.852 us |
| UrlEncodeAWS_SymbolCompare | 403.22 us | 3.371 us | 3.153 us |
|  UrlEncodeMDW_SymbolSearch | 251.31 us | 1.342 us | 1.255 us |
| UrlEncodeMDW_SymbolCompare |  40.70 us | 0.599 us | 0.560 us |



.NET 6.0.21:

|                     Method |      Mean |    Error |   StdDev |
|--------------------------- |----------:|---------:|---------:|
|               UrlEncodeAWS | 165.88 us | 1.545 us | 1.446 us |
|               UrlEncodeMDW |  50.43 us | 0.371 us | 0.347 us |
|       UrlEncodeMDWWithMask |  57.53 us | 0.539 us | 0.450 us |
|  UrlEncodeAWS_SymbolSearch | 168.18 us | 1.406 us | 1.246 us |
| UrlEncodeAWS_SymbolCompare | 137.46 us | 1.442 us | 1.349 us |
|  UrlEncodeMDW_SymbolSearch |  58.29 us | 0.311 us | 0.291 us |
| UrlEncodeMDW_SymbolCompare |  29.00 us | 0.211 us | 0.187 us |

WithMask vs the unadorned without mask are just measurement noise issues. Over multiple runs, they'll flip which one is longer.

(SymbolSearch vs SymbolCompare is looking at the cost of `unreservedChars.IndexOf(symbol)` vs character range checks, but that change probably has a higher long term cost. This is the `COMString::IndexOfChar` referenced by the earlier profile "analysis".)

boblodgett commented on sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:
```diff
@@ -1081,12 +1081,31 @@ public static string UrlEncode(int rfcNumber, string data, bool path)
+                    // Break apart the byte into two four-bit components and
+                    // then convert each into their hexadecimal equivalent.
+                    byte b = (byte)symbol;
+                    int hinibble = b >> 4;
+                    int lonibble = b & 0xF;
```
Can you name this `hiNibble` and `loNibble`?

boblodgett commented on sdk/src/Core/Amazon.Util/AWSSDKUtils.cs:
```diff
@@ -1081,12 +1081,31 @@ public static string UrlEncode(int rfcNumber, string data, bool path)
             return encoded.ToString();
         }
+
+        private static char ToUpperHexit(int value)
```
Can you rename to `ToUpperHex`?

boblodgett commented:
I have added this change to our release queue so we should be able to get this merged in during our next out-of-band release. We will update here once we get it merged in. Thank you for the PR!

mwinterb commented:
> I have added this change to our release queue so we should be able to get this merged in during our next out-of-band release. We will update here once we get it merged in. Thank you for the PR!

Excellent, thanks all.

normj commented:
The PR has been rebased into master and released as part of version 3.7.202.16 of AWSSDK.Core
